refactor codes of inserting cast

mingmingtasd opened the pull request:
Fix #87 

1. refactor codes for inserting cast node 
2. support logical not and fix bugs for all logical operators

PTAL, thanks!

shiyi9801 commented on services/webnn/ort/graph_builder_ort.cc:
```diff
@@ -114,12 +114,19 @@ base::unexpected<mojom::ErrorPtr> NewUnknownError(std::string message) {
+// TODO(https://github.com/shiyi9801/chromium/issues/63): Make name generation
+// more robust. Inserted operations should also have a unique name, so here
+// they're named by their ids for now.
+std::string GetInsertedOperationName(uint64_t operation_id) {
```
Instead of using `GetInsertedOperationName()` and `next_operation_count_`, please create a label for each inserted operation (e.g. `node_name + "_cast"`) and then call `GetNodeName(std::string_view label)` to ensure there is no duplicate operation names, also to ensure we get more useful error log.

shiyi9801 replied:
`node_name + "_cast"` -> `label + "_cast"`

shiyi9801 commented on services/webnn/ort/graph_builder_ort.cc:
```diff
@@ -321,6 +329,35 @@ std::string GraphBuilderOrt::CreateScalarInitializer(const DataType& value) {
+  input_name = output_name;
+}
+
+void GraphBuilderOrt::AppendCast(const std::string& input_name,
```
Please use `std::string_view` instead of `const std::string&` here.

shiyi9801 commented on services/webnn/ort/graph_builder_ort.cc:
```diff
@@ -114,12 +114,19 @@ base::unexpected<mojom::ErrorPtr> NewUnknownError(std::string message) {
 // TODO(https://github.com/shiyi9801/chromium/issues/63): Make name generation
-// more robust. Inserted operands should also have a unique id, so here they're
-// named by their ids for now.
+// more robust. Inserted operands should also have a unique name, so here
```
Here the comment meant each operand (including the inserted ones) is **guaranteed** to have a unique id, and that's why they can be named by their ids.

shiyi9801 reviewed (commented):
Thanks for fixing the issue of logical ops!

huningxin commented on services/webnn/ort/graph_builder_ort.h:
```diff
@@ -145,6 +146,20 @@ class GraphBuilderOrt {
+  // target `to_data_type`. The `input_name` specifies the cast operation's
+  // input (the operation's output), and the `output_name` specifies the cast
+  // operation's output.
+  void AppendCast(const std::string& input_name,
```
Use `std::string_view`?

mingmingtasd commented on services/webnn/ort/graph_builder_ort.cc:
```diff
@@ -321,6 +329,35 @@ std::string GraphBuilderOrt::CreateScalarInitializer(const DataType& value) {
       /*shape=*/{}, base::span_from_ref(value));
 }
 
+void GraphBuilderOrt::PrependCast(std::string& input_name,
```
I intend to make the codes more friendly and easier for callers, so I prefer keeping the `PrependCast` and `AppendCast` methods. And I also add `#define ADD_CAST_NODE` to reuse the code. PTAL.

mingmingtasd commented:
Addressed your comments, PTAL, thanks!
Sorry for that I have to force push because after rebasing, I can push my new commit. All new changes are in the new commit.

huningxin commented:
> Sorry for that I have to force push because after rebasing, I can't push my new commit. All new changes are in the new commit

No worries, thanks for making the new commit.

huningxin commented on services/webnn/ort/graph_builder_ort.cc:
```diff
@@ -315,9 +313,8 @@ std::string GraphBuilderOrt::CreateInitializer(
 
   CHECK(result_->id_to_operand_info
-            .try_emplace(next_operand_id_, std::move(operand_info))
+            .try_emplace(next_operand_id, std::move(operand_info))
```
Where is `result_->id_to_operand_info` used? By whom? I know this is an old issue. But I feel confused when reviewing the usage of `next_operand_id`. Does ONNX graph only need to use the names at all?

shiyi9801 replied:
`result_->id_to_operand_info` was intended to be used by Dispatch to get input/output names, but at current stage it's useless and looks like it won't be used at all. I can upload a PR to remove `result_->id_to_operand_info` as a follow-up.

huningxin reviewed (approved):
LGTM with a question

huningxin commented:
You may need to rebase again.

shiyi9801 reviewed (approved):
👍

mingmingtasd commented:
> You may need to rebase again.

Rebased again for the resample2d change.
